aiohttp/signals.py

```diff
@@ -1,95 +1,32 @@
-import asyncio
-from itertools import count
-
 from aiohttp.frozenlist import FrozenList
 
 
-class BaseSignal(FrozenList):
-
-    __slots__ = ()
-
-    async def _send(self, *args, **kwargs):
-        for receiver in self:
-            res = receiver(*args, **kwargs)
-            if asyncio.iscoroutine(res) or asyncio.isfuture(res):
-                await res
-
-
-class Signal(BaseSignal):
+class Signal(FrozenList):
     """Coroutine-based signal implementation.
 
     To connect a callback to a signal, use any list method.
 
-    Signals are fired using the :meth:`send` coroutine, which takes named
+    Signals are fired using the send() coroutine, which takes named
     arguments.
     """
 
-    __slots__ = ('_app', '_name', '_pre', '_post')
+    __slots__ = ('_owner',)
 
-    def __init__(self, app):
+    def __init__(self, owner):
         super().__init__()
-        self._app = app
-        klass = self.__class__
-        self._name = klass.__module__ + ':' + klass.__qualname__
-        self._pre = app.on_pre_signal
-        self._post = app.on_post_signal
+        self._owner = owner
+
+    def __repr__(self):
+        return '<Signal owner={}, frozen={}, {!r}>'.format(self._owner,
+                                                           self.frozen,
+                                                           list(self))
 
     async def send(self, *args, **kwargs):
         """
         Sends data to all registered receivers.
         """
-        if self:
-            ordinal = None
-            debug = self._app._debug
-            if debug:
-                ordinal = self._pre.ordinal()
-                await self._pre.send(
-                    ordinal, self._name, *args, **kwargs)
-            await self._send(*args, **kwargs)
-            if debug:
-                await self._post.send(
-                    ordinal, self._name, *args, **kwargs)
-
+        if not self.frozen:
+            raise RuntimeError("Cannot send non-frozen signal.")
 
-class FuncSignal(BaseSignal):
-    """Callback-based signal implementation.
-
-    To connect a callback to a signal, use any list method.
-
-    Signals are fired using the :meth:`send` method, which takes named
-    arguments.
-    """
-
-    __slots__ = ()
-
-    def send(self, *args, **kwargs):
-        """
-        Sends data to all registered receivers.
-        """
         for receiver in self:
-            receiver(*args, **kwargs)
-
-
-class DebugSignal(BaseSignal):
-
-    __slots__ = ()
-
-    async def send(self, ordinal, name, *args, **kwargs):
-        await self._send(ordinal, name, *args, **kwargs)
-
-
-class PreSignal(DebugSignal):
-
-    __slots__ = ('_counter',)
-
-    def __init__(self):
-        super().__init__()
-        self._counter = count(1)
-
-    def ordinal(self):
-        return next(self._counter)
-
-
-class PostSignal(DebugSignal):
-
-    __slots__ = ()
+            await receiver(*args, **kwargs)
```

Comment:
How about to add more friendly exception if plain func is going to be registered as handler? I guess now such code will fail with cryptic `TypeError: object NoneType can't be used in 'await' expression` or similar.

Reply:
Do you suggest catching `TypeError` and raise a new `TypeError` with more informative exception message?

Reply:
As an option we could check ll callbacks on freezing stage.
It prevents registering a regular function which returns a future but it's fine I guess.

Reply:
Yes for TypeError and yes, better check this early. On append call if possible. The stacktrace will help to find that bad signal usage with easy.

Reply:
FrozenList can be modified by too many ways, `.append` is not the only one.
I think `.freeze()` is a good compromise.

Reply:
It's a pity. Ok then if error message would contains function name which is not async.

Reply:
Well, but I see you point.
Maybe we need specialized `CheckedFrozenList` class with registered callback? It could improve UX slightly.
Please let me merge the PR as is and make a new issue.
I don't want to block @pfreixes with his client tracing PR.

Reply:
Ok. Let's do that.